### PR TITLE
[DO NOT MERGE][handle] Migrate System.Globalization icalls

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

File: mono/metadata/handle.c [hunk 1]
```diff
@@ -203,25 +178,30 @@ arena_scan (gpointer addr, MonoGCMarkFunc mark_func, gpointer gc_data)
...
 {
-#ifdef HAVE_SGEN_GC
 	mono_gc_deregister_root ((char*) arena_stack);
-#endif
 }
 
+MonoHandleArena*
+mono_handle_arena_current (void)
+{
+	g_assert_not_reached ();
```
Comment (lambdageek) [inline mono/metadata/handle.c]: Depends on #2398 (or something like it).

**Comment by ludovic-henry** [comment]
@kumpera: you can find your previous comments here https://github.com/ludovic-henry/mono/compare/bb6761542e50f568b6d2259aae437cc4088ae5d1...a09ab3e

File: mono/metadata/handle.h [hunk 2]
```diff
@@ -32,15 +24,33 @@ typedef MonoHandleStorage* MonoHandle;
...
-struct _MonoHandleStorage {
-	MonoObject *obj;
-};
+typedef struct {
+	MonoObject *__private_obj;
+} MonoHandleStorage;
+
+typedef MonoHandleStorage* MonoHandle;
+
+#include "handle-arena.h"
```
Comment (lambdageek) [inline mono/metadata/handle.h]: You can do
```
  /* here */
  #include "handle-arena.h"
  struct _MonoHandleStorage {
     MonoObject *__private_obj;
  }
  typedef struct _MonoHandleStorage MonoHandleStorage;
  typedef MonoHandleStorage* MonoHandle;

  /* in handle-arena.h */
  typedef struct _MonoHandleStorage MonoHandleStorage;
  typedef MonoHandleStorage* MonoHandle;
```

The stuff in `handle-arena.h` doesn't actually need to see the definition of `struct _MonoHandleStorage` in order to define a typedef to it.  C is funny that way.

Comment (ludovic-henry) [inline mono/metadata/handle.h]: Wouldn't it be simpler to inline handle-arena.h in handle.h? We have to expose both anyway.

Comment (lambdageek) [inline mono/metadata/handle.h]: Up to you. I don't have a strong preference either way.

(My weak preference is for "headers in a separate 'include' directory with nested subdirectories, one thing per header, roll-up headers for client convenience in base directories", but that's not what Mono looks like, generally, so it's not particularly relevant)

**Comment by lambdageek** [comment]
@ludovic-henry : Cherry-pick [this one line](https://github.com/lambdageek/mono/commit/3f91fc09c356b9491d8ed90105a446d40365de20#diff-45ebb3d3ae14abc6bf0815c343fe0cd4R19) so that everything compiles in checked builds, too.

**Comment by lambdageek** [comment]
With the handle API as it is today, the handle API should not be used used on code paths that unwind the native stack, the handle arena stack will be left in a corrupted state.

Is there some way we can modify the throwing code to roll back the arena stack?

**Comment by kumpera** [comment]
The unwinder is a pretty damn complicated piece of code and triggering of unwinding runtime frame is a bad behavior we currently have that must go.

**Comment by lambdageek** [comment]
LGTM

**Comment by akoeplinger** [comment]
build

**Comment by akoeplinger** [comment]
build

**Comment by kumpera** [comment]
Hey guys,

Here's my review of the changes:

The following patches are great and can go in:

more ok
f9b0187
fc7ff95
a61a721
999d6a9
fefb2c6
5115814
11b8401
783a13d
81de6b2
ea53116
cc309fd
1f88a3d

The following patch requires the documentation on Coop to be updated:
a61a721 (here http://www.mono-project.com/docs/advanced/runtime/docs/coop-suspend/)

The following commits are not ok for the following reasons:
They are forcing a slow path onto common code today.
They are forcing handles all over the place before it's ready. It still *doesn't work with boehm*.
The approach is wrong. We should not push handles into gc specific code, we'll regret that
later. It will be better if we lift OOM concerns from the gc specific code and handle them in metadata.

5c80c80
5ad6caa
aadbf4a


The following commit has the following issues:
It's moving code around for no good reason.
It's renaming functions for no good reason.
It's reformatting code for no good reason.
Just the above 3 are enough to throw it out of consideration, but there are some issues with the code.
The icalls are ignoring all discussion on entry entry protocols we discussed before.
Locals are being declared in the worst possible way. They look fugly.
7757891

**Pull request closed without merging**